File: Euler/factorialDigitSum/fds.cpp
#include <iostream>
#include <vector>
#include <cstdlib>

using namespace std;

unsigned long long dynFactorial(unsigned long long num, vector<unsigned long long> *fs){

  //Base Case
  if(num == 0){
    return 1;
  }

  //Populate current slot if not done.
  if(fs->at(num-1) == 0){
    fs->at(num-1) = dynFactorial(num - 1,fs) * num;
    return fs->at(num-1);
  }else{
    return fs->at(num-1);
  }


}

int main(int argc, char** argv){

  if(argc != 2){
    cerr << "Usage: " << argv[0] << " <number> "<< endl;
    return -1;
  }

  unsigned long long num = strtoul(argv[1],NULL,0);
  vector<unsigned long long> factStorage;
  for(unsigned long long i = 0; i < num; i++){
    factStorage.push_back(0);
  }

  cout << "The factorial of " << num << " is " << dynFactorial(num,&factStorage) << endl;


  return 0;
}

File: Euler/factorialDigitSum/fds.java
import java.math.*;
import java.util.ArrayList;

class fds{

  public static ArrayList<BigInteger> fs = new ArrayList<BigInteger>();

  public static BigInteger dynFactorial(int num){
    if(num == 0){
      return new BigInteger("1");
    }

    //System.out.println("Working on number " + num + " current state is " + fs.get(num - 1));
    
    if(fs.get(num - 1).intValue() == 0){
      fs.set(num - 1,dynFactorial(num - 1).multiply(new BigInteger(String.valueOf(num))));
    //  System.out.println("Working on number " + num + " changed state is " + fs.get(num - 1));
    }
  
    return fs.get(num-1);
  }

  public static void main(String[] args){
  
    if(args.length != 1){
      System.err.println("Usage: java fds <Number>");
      return;
    }
    int num = Integer.parseInt(args[0]);  
    for(int i = 0; i < num; i++){
      fs.add(new BigInteger("0"));
    }
    System.out.println("The length of the list is " + fs.size()); 
    BigInteger dyNum = dynFactorial(num);
    System.out.println("The factorial for number " + num + " is " + dyNum);
 
    int sum = 0;
    for(int i = 0; i < dyNum.toString().length(); i++){
      sum += new Integer(Character.getNumericValue(dyNum.toString().charAt(i)));
    }

    System.out.println("The sum of digits of " + dyNum + " is " + sum);
  }
}
